### PR TITLE
Add anyref feature and type

### PR DESCRIPTION
**Pull request by dcodeIO**
Adds a custom `--enable-anyref` feature enabling just the `anyref` type on top of reference types that allows us to test subtyping relationship (with `externref`, `funcref` and `exnref`) without having to enable the full set of GC features. Re-enables previously disabled tests.

### CODE REVIEW

File: src/wasm/wasm-type.cpp
Comment (dcodeIO) [inline src/wasm/wasm-type.cpp:547]: Fixing this looks like it requires changing the signature of this function to `getLeastUpperBound(Type a, Type b, FeatureSet features)`, with call sites aware of enabled features passing these in.

Only works this way currently because we are not generating, for example, an `(if $condition $funcref $externref)` anywhere, but an API user can do so. Do you prefer to tackle this issue in this PR?

File: src/tools/fuzzing.h
Comment (dcodeIO) [inline src/tools/fuzzing.h:328]: Found that at most places, similar checks are not nested inside `if (hasReferenceTypes)`. Not an urgent problem, but I'd imagine that this is going to lead to obscure issues where a user specifies `--enable-anyref` but not `--enable-reference-types` eventually.

One way to tackle this might be to implicitly enable reference types on `setAnyref(true)` respectively `setExceptionHandling(true)`, and implicitly disable anyref and exception handling on `setReferenceTypes(false)`?

**Comment by dcodeIO** [comment]
Some early fuzzing:

```
Invocations so far:
   FuzzExec: 5875
   CompareVMs: 1441
   CheckDeterminism: 510
   Wasm2JS: 1351
   Asyncify: 1412

ITERATION: 6816
```

File: src/wasm/wasm-type.cpp
Comment (tlively) [reply src/wasm/wasm-type.cpp:547]: Unless I'm missing some use of this function, I believe it will only incorrectly return `anyref` for invalid code. As long as validation will separately catch that invalidity, I'm tempted to not do anything with this issue besides document it. WDYT?

File: src/tools/fuzzing.h
Comment (tlively) [reply src/tools/fuzzing.h:328]: Yeah this is a tricky issue. So far we've avoided having any implicit dependencies between features to keep the mental model as simple and predictable as possible. One possibility would be to make it a validation error to have `anyref` enabled without having reference types enabled. That's a little more annoying for users, but on the other hand it keeps the features simple and there should not be real users of the `--enable-anyref` feature anyway.

File: src/wasm/wasm-type.cpp
Comment (tlively) [reply src/wasm/wasm-type.cpp:547]: Ah, it looks like this PR doesn't make any changes to validation yet. It would be good to fix that.

**Comment by tlively** [comment]
This LGTM once those two conversations are resolved 👍

File: src/wasm/wasm-type.cpp
Comment (dcodeIO) [reply src/wasm/wasm-type.cpp:547]: Have a feeling that this will resurface eventually, but since the fuzzer isn't complaining I guess it is fine for now. Is the comment ok, or would you prefer a little more detail?

Comment (dcodeIO) [reply src/wasm/wasm-type.cpp:547]: Oh, going to check!

**Comment by aheejin** [comment]
Thanks for doing this. I quickly glanced over a few re-enabled subtype tests and I am a little confused; some tests seem to have their supertype-subtype order reversed compared to the original tests; and it is hard to check that because, disabled tests in #3084 already have their types changed.

The one I discovered was done by comparing my local code (without changes from #3084) and this PR. I just started looking at this code just now, but can you hold off on landing this?

**Comment by dcodeIO** [comment]
Yeah, these changes can be confusing. Previously we had

```
nullref <: externref | funcref | exnref
```

with `nullref` being a common subtype, but now we have

```
externref | funcref | exnref <: anyref
```

wit `anyref` being a common supertype, effectively reversing the order.

File: src/wasm/wasm-type.cpp
Comment (dcodeIO) [reply src/wasm/wasm-type.cpp:547]: On a first glimpse it looks like the validator uses `Type::isSubType` for its checks. This then leads to a similar issue like this one for `getLeastUpperBound`, but for `isSubType` not knowing the enabled features I guess, where the check in `isSubType` is `right == Type::anyref` which can only happen in valid code if `anyref` is enabled.

Can you point me into the right direction what else needs to be updated in the validator?

**Comment by aheejin** [comment]
Oh sorry. You're right, and I was aware of that. I thought the case I meant was not that and really had a relationship reversed, but I just realized that the line wrapping changed and I was confused by that. Anyway, the case I discovered didn't have any problem, and I trust you on other tests. :) Sorry for the confusion and thanks!

File: src/wasm/wasm-type.cpp
Comment (dcodeIO) [reply src/wasm/wasm-type.cpp:547]: On the other hand, there will be a(nother) validation error anyway if `anyref` is used but not enabled, via `Type::getFeatures()`.

File: src/tools/fuzzing.h
Comment (aheejin) [reply src/tools/fuzzing.h:328]: I had the same problem for EH, and I've been doing this kind of things for EH features, to avoid the feature dependences.. https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/tools/fuzzing.h#L2659-L2660

File: src/wasm/literal.cpp
Comment (aheejin) [inline src/wasm/literal.cpp:395]: In which case do we need non-null `anyref` and `externref` literals?

Comment (dcodeIO) [reply src/wasm/literal.cpp:395]: This came up earlier in that these are currently not supported by `Literal`, which only can represent `null`s of any reference type, but Thomas mentioned that we'd want to support these eventually, analogous to `ExceptionPackage` for `exnref`, so the interpreter can run code by representing non-null `anyref` etc. at least abstractly.

File: src/wasm/wasm-type.cpp
Comment (aheejin) [reply src/wasm/wasm-type.cpp:547]: If this ends up generating `anyref` in other programs that don't use or enable `anyref`, eventually it will be caught by the validator, and that basically means one of Binaryen passes have a bug. So I think it's OK to keep this simple and not include types for now, but erroring out here might help Binaryen authors debug their code a little, or something.

I think `isSubType` will be OK, because it doesn't create `anyref` when it isn't enabled.

File: src/tools/fuzzing.h
Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: Added a feature validation step, but naturally this now breaks the fuzzer because it doesn't know of implied features. Any suggestions?

Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: Ended up with https://github.com/WebAssembly/binaryen/pull/3109/commits/fb1a789b0efdc88c6522e6986d34f2e101da77a4 for now

Comment (aheejin) [reply src/tools/fuzzing.h:328]: I don't think the fuzzer will generate types that does not respect dependences. For example, we [check](https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/tools/fuzzing.h#L2659-L2660) for both `FeatureSet::ExceptionHandling` and `FeatureSet::ReferenceTypes` whenever we try to add `exnref` or other EH instructions (By the way I just realized my PR for generating EH instructions for the fuzzer hasn't landed and is pending for months, but anyway). So I guess the fuzzer error you've seen was just complaining about the feature combination itself, right?

So I'm wondering, why do we need to impose the dependences in validator? Unless giving `--enable-exception-handling` without `--enable-reference-types` itself in the command line of a VM is an error, I don't think we should treat it as an error.

Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: > So I guess the fuzzer error you've seen was just complaining about the feature combination itself, right?

Correct, just due to adding the feature validation to the validator and the fuzzer picking random feature opts.

> So I'm wondering, why do we need to impose the dependences in validator?

The initial concern was that at most places we have code like

```
if (hasReferenceTypes) {
  add(funcref)
  add(externref)
}
if (hasExceptionHandling) {
  add(exnref);
}
if (hasAnyref) {
  add(anyref);
}
```

instead of

```
if (hasReferenceTypes) {
  add(funcref)
  add(externref)
  if (hasExceptionHandling) {
    add(exnref);
  }
  if (hasAnyref) {
    add(anyref);
  }
}
```

so it seems possible to yield an `exnref` or `anyref` without reference types being enabled, or to otherwise get false positives, and a validation check lets us keep the code as-is.

Alternatives:
* Refactor all relevant code locations to check for all required features or
* Upon enabling exception handling or anyref, also enable reference types implicitly / upon disabling reference types, also disable exception handling and anyref implicitly

Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: From a user's perspective:

* If making missing required features validation errors: "Oh, I see"
* If refactoring all code locations: "Why isn't this option picked up, even though it's specified?"
* If enabling features implicitly: "Why does my binary suddenly utilize reference types, even though it isn't specified?"

From our perspective:

* If making missing required features validation errors: -
* If refactoring all code locations: "Look, a new issue"
* If enabling features implicitly: "Look, a new issue"

File: src/wasm/wasm-type.cpp
Comment (tlively) [reply src/wasm/wasm-type.cpp:547]: > On the other hand, there will be a(nother) validation error anyway if `anyref` is used but not enabled, via `Type::getFeatures()`.

Aha, I had forgotten that this wouldn't require any changes in the validator itself. I agree with @aheejin that this should be fine, then. We can probably replace the `FIXME` with a comment pointing out the issue but explaining that it will only happen with invalid code.

File: scripts/fuzz_opt.py
Comment (tlively) [inline scripts/fuzz_opt.py:116]:
```suggestion
                if possible in IMPLIED_FEATURE_OPTS:
                    FEATURE_OPTS.extend(IMPLIED_FEATURE_OPTS[possible])
```

File: src/tools/fuzzing.h
Comment (aheejin) [reply src/tools/fuzzing.h:328]: > The initial concern was that at most places we have code like
> 
> ```
> if (hasReferenceTypes) {
>   add(funcref)
>   add(externref)
> }
> if (hasExceptionHandling) {
>   add(exnref);
> }
> if (hasAnyref) {
>   add(anyref);
> }
> ```
> 
> instead of
> 
> ```
> if (hasReferenceTypes) {
>   add(funcref)
>   add(externref)
>   if (hasExceptionHandling) {
>     add(exnref);
>   }
>   if (hasAnyref) {
>     add(anyref);
>   }
> }
> ```
> 
> so it seems possible to yield an `exnref` or `anyref` without reference types being enabled, or to otherwise get false positives, and a validation check lets us keep the code as-is.

I wonder where we do something like the first? The only place I can think of that _creates_ `exnref` or `anyref` is the fuzzer, and I think it handles them correctly (or you handled them correctly in this PR). Validation or checking wouldn't be as problematic as creating them because they don't bother with `anyref` of `exnref` when they don't exist. I could be mistaken; if so please let me know.

Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: Checking again, there is one place where we are doing the former

https://github.com/WebAssembly/binaryen/blob/916ce6f1a9f7c85102a8c69f593b301c8df5d19d/src/passes/InstrumentLocals.cpp#L170-L193

and I somehow assumed that this is some sort of convention, even though it probably isn't. Would suggest to move the `hasExceptionHandling` into `hasReferenceTypes` there as well so there can't be instrumented `get_exnref`/`set_exnref` functions taking an `exnref` type if reference types is not enabled, but exception handling is. Validating separately still seems fine to me, though, for clarity.

Comment (dcodeIO) [reply src/tools/fuzzing.h:328]: (Haven't checked other locations where just `hasExceptionHandling` is checked, but `hasReferenceTypes` is not)

Comment (aheejin) [reply src/tools/fuzzing.h:328]: Thanks for finding this. That is a special pass that generates imports, and I agree that moving `hasExceptionHandling` condition into `hasReferenceTypes` is the right call.. Can you add that to this PR if possible?

But I don't think there are many, if at all, passes that generates `anyref` of `exnref` out of thin air. (Except for the fuzzer, but it is not a pass) One function that can generate them and can be called from Binaryen passes is `getLeastUpperBound`, as you pointed out, which you handled correctly in this PR I believe. So I don't think there will be mass refactoring necessary. And I'd rather not implicitly enable dependent features; it can be an unexpected behavior for users.

I just checked what V8 and wabt does for feature dependencies, and actually they are both doing the implication thing, so if you only enable EH that automatically enables reference types too, even if you don't explicitly enable it. Maybe we should consider doing a similar thing in future.

**Review by aheejin — APPROVED** [review]
LGTM, and thanks!

**Comment by dcodeIO** [comment]
Please let me know whether the recent series of commits sufficiently addresses your concerns or if I misunderstood something :)

File: CHANGELOG.md
Comment (aheejin) [inline CHANGELOG.md:21]: This has been the case before (that you have to specify dependent features too), so nothing has changed here. If we later add feature implying feature then we need to write something here.

But it'd be worth noting that `--enable-anyref` is added, and what it does.

File: src/wasm/wasm-type.cpp
Comment (aheejin) [inline src/wasm/wasm-type.cpp:551]: - I think if user's input code was verified and accepted, this cannot be a problem in the user's end. If user's code had a problem, it should've been rejected when reading because it was using not enabled features.

> If it turns out that it can happen otherwise, this method would need to take the enabled feature set as an additional argument at all call sites.
- I don't think it can happen otherwise, but maybe it can, and I'm missing something. But if that happens we can fix it; I'm not sure we need to express uncertainties in the comments..
- I think feature set implication (so that if you enable only EH, reference types is also enabled) is something that's good to have regardless of whether this can happen or not. I was skeptical, but apparently other tools and VMs are doing it too. Of course that will be a future task (for anyone interested) and I'm not suggesting we should implement that here.

File: src/tools/fuzzing.h
Comment (aheejin) [inline src/tools/fuzzing.h:322]: I'm not too opinionated about this, but the fuzzer still does not assume feature dependencies have been completely resolved; it checks that here and there. How about restoring this condition and maybe we can remove it when we add feature implication feature.

File: CHANGELOG.md
Comment (dcodeIO) [reply CHANGELOG.md:21]: Added this because it is a validation error now (as of the current state of this PR), but wasn't before. Previously one would only get a validation error further down the road if `--enable-exception-handling` was specified, an `exnref` emitted somehow but `--enable-reference-types` omitted.

File: src/wasm/wasm-type.cpp
Comment (dcodeIO) [reply src/wasm/wasm-type.cpp:551]: Oh, with "the user's end" I meant a generator like AssemblyScript, using the C or JS APIs. A generator can freely emit `anyref`, or `select`s with differing reftype arms even if a feature is not enabled, which would be a bug on their end.

In general I still feel like this should be fixed properly instead, and I'm having a hard time coming up with a comment here that doesn't sound like "lol, this will most likely break". May I ask for a concrete suggestion I can copy and paste?

File: src/tools/fuzzing.h
Comment (dcodeIO) [reply src/tools/fuzzing.h:322]: That's exactly why I assumed that if `anyref` is present, the fuzzer wouldn't have emitted it earlier if reference types aren't enabled, hence removing the `assert`. Yet, `exnref` isn't a subtype if exception handling isn't enabled, so the fuzzer shouldn't emit it in this case, and I thought this change would be an improvement. May I ask for a concrete suggestion on what to change it to?

File: src/wasm/wasm-type.cpp
Comment (tlively) [reply src/wasm/wasm-type.cpp:551]:
```
The LUB of two different reference types is anyref, which may or may not be a valid type depending on whether the `anyref` feature is enabled. When `anyref` is disabled, it is possible for the finalization of invalid code to introduce a use of anyref via this function, but that is not a problem because it will be caught and rejected by validation.
```

File: src/tools/fuzzing.h
Comment (dcodeIO) [reply src/tools/fuzzing.h]: Gone ahead and restored the condition. Hope it's correct now :)

File: CHANGELOG.md
Comment (dcodeIO) [reply CHANGELOG.md]: Split the entry into two parts, with one mentioning the new feature and one mentioning that not enabling implied features is a validation error now.

**Comment by dcodeIO** [comment]
Feel free to merge if it looks ok to you now. Not quite confident that your concerns are addressed. Meanwhile looking for something useful to do, now eyeballing the name section, i.e. module and local names.

**Review by tlively — APPROVED** [review]

File: CHANGELOG.md
Comment (aheejin) [reply CHANGELOG.md]: Oh sorry, I misinterpreted the comment. I actually forgot that this PR makes it a validation error if dependent flags aren't specified. I was not in favor of that, but maybe we should do feature implication later.

**Pull request merged**
